tests/unit/scenario_/database_relations/test_database_relations_breaking.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Test app status and relation databags"""
+
+import ops
+import pytest
+import scenario
+
+import machine_charm
+
+from ..wrapper import Relation
+from . import combinations
+
+
+def output_state(
+    *, relations: list[Relation | scenario.Relation], event: scenario.Event
+) -> scenario.State:
+    for index, relation in enumerate(relations):
+        if isinstance(relation, Relation):
+            relations[index] = relation.freeze()
+    relations: list[scenario.Relation]
+    context = scenario.Context(machine_charm.MachineSubordinateRouterCharm)
+    input_state = scenario.State(
+        relations=relations,
+        leader=True,
+    )
+    return context.run(event, input_state)
+
+
+@pytest.mark.parametrize("complete_provides_s", combinations.complete_provides(1, 3))
+def test_breaking_requires_and_complete_provides(complete_requires, complete_provides_s):
+    for provides in complete_provides_s:
+        provides.local_app_data = {
+            "database": "foobar",
+            "endpoints": "file:///var/snap/charmed-mysql/common/run/mysqlrouter/mysql.sock",
+            "read-only-endpoints": "file:///var/snap/charmed-mysql/common/run/mysqlrouter/mysqlro.sock",
+            "username": "foouser",
+            "password": "foobar",
+        }
+    complete_requires = complete_requires.freeze()
+    state = output_state(
+        relations=[complete_requires, *complete_provides_s], event=complete_requires.broken_event
+    )
+
+    assert state.app_status == ops.BlockedStatus("Missing relation: backend-database")
+    for index, provides in enumerate(complete_provides_s, 1):
+        assert state.relations[index].local_app_data == {}
+
+
+@pytest.mark.parametrize("complete_provides_s", combinations.complete_provides(1, 3))
+def test_complete_requires_and_breaking_provides(complete_requires, complete_provides_s):
+    for provides in complete_provides_s:
+        provides.local_app_data = {
+            "database": "foobar",
+            "endpoints": "file:///var/snap/charmed-mysql/common/run/mysqlrouter/mysql.sock",
+            "read-only-endpoints": "file:///var/snap/charmed-mysql/common/run/mysqlrouter/mysqlro.sock",
+            "username": "foouser",
+            "password": "foobar",
+        }
+    # Needed to access `.broken_event`
+    complete_provides_s = [relation.freeze() for relation in complete_provides_s]
+
+    state = output_state(
+        relations=[complete_requires, *complete_provides_s],
+        event=complete_provides_s[-1].broken_event,
+    )
+    if len(complete_provides_s) == 1:
+        assert state.app_status == ops.BlockedStatus("Missing relation: database")
+    else:
+        assert state.app_status == ops.ActiveStatus()
+    assert state.relations[-1].local_app_data == {}
+    complete_provides_s.pop()
+    for index, provides in enumerate(complete_provides_s, 1):
+        assert state.relations[index].local_app_data == {
```

Comment:
take a look at `State.get_relations`, maybe it can be handy for filtering out what relations you want to look at.

I spent some time thinking about how it'd be handy to expose the relations, but I wasn't sure.

would a `get_relation_by_id` be better?

Reply:
yeah precisely. I had this issue: https://github.com/canonical/ops-scenario/issues/20 to track ideas and proposals around this use case

Reply:
something like `obj_out = state_out.get_me_corresponding_object(obj_in)`